index/index.py

```diff
@@ -193,35 +196,58 @@ def find_ref(search_term, max_l_dist=2, num_results=5):
         print("No matches found")
 
 
-def add_urn(lemma, lemma_type, urn=None, overwrite=False):
-    """Add a urn to a lemma of type 'author' or 'work'"""
+def add_metadata(md_type, lemma, lemma_type, md_value=None, overwrite=False):
+    """Add metadata to a lemma of type 'author' or 'work'"""
+
+    # Validation: metadata type is correct
+    if not md_type in METADATA_TYPES:
+        raise click.BadParameter(f"'{md_type} is not a valid metadata type.'")
 
     # Open index file
     with open(INDEX_FILE) as f:
         index = yaml.load(f)
 
+    # Validation: {md_type: md_value} is not already present in index.
+    for l in index:
+        # If the lemma already has this value, other logic takes care of it.
+        if l == str(lemma):
+            continue
+
+        lemma_dict = index[l]
+        try:
+            lemma_md_dict = lemma_dict["metadata"]
+            lemma_md_value = lemma_md_dict[str(md_type)]
+        except KeyError:
+            continue
+        else:
+            if lemma_md_value == str(md_value):
+                raise click.BadParameter(f"{md_value} is already defined as {md_type} for {l}. {md_type} must be unique.""")
+            else:
+                continue
+
+    
     # Validation: lemma exists
-    if lemma not in index:
+    try:
+        lemma_dict = index[lemma]
+    except KeyError:
         raise click.BadParameter(f"Lemma '{lemma}' not found in index.")
 
     # Validation: lemma type is correct
     if lemma_type not in LEMMA_TYPES:
         raise click.BadParameter(f"'{lemma_type}' is not a valid lemma type.")
 
-    lemma_dict = index[lemma]
-
     # Validation: lemma in index is of same type as lemma_type
     if lemma_type != lemma_dict.get("type"):
         raise click.BadParameter(
             f"Lemma is in the index as '{lemma_dict.get('type')}', not as '{lemma_type}'."
         )
 
-    if urn:
+    if md_type == 'cts_urn':
```

Comment:
Mooie use case voor de nieuwe [`match case`](https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching) pattern!

Reply:
Hm, vreemd dat dat niet werkt. Die `str(...)` is volgens mij overbodig, maar zou het niet moeten breken. Laat maar zitten als je er niet uit komt..